File: rerun_py/rerun_sdk/rerun/datatypes/rotation3d_ext.py
from __future__ import annotations

from typing import TYPE_CHECKING

import numpy as np

if TYPE_CHECKING:
    from . import Quaternion, Rotation3D, Rotation3DLike, RotationAxisAngle


class Rotation3DExt:
    """Extension for [Rotation3D][rerun.datatypes.Rotation3D]."""

    @staticmethod
    def identity() -> Rotation3D:
        from . import Quaternion, Rotation3D

        return Rotation3D(Quaternion.identity())

    @staticmethod
    def inner__field_converter_override(
        data: Rotation3DLike,
    ) -> Quaternion | RotationAxisAngle:
        from . import Quaternion, Rotation3D, RotationAxisAngle

        if isinstance(data, Rotation3D):
            return data.inner
        elif isinstance(data, (Quaternion, RotationAxisAngle)):
            return data
        else:
            return Quaternion(xyzw=np.array(data))

File: rerun_cpp/src/rerun/components/rotation3d_ext.cpp
#include "rotation3d.hpp"

// Uncomment for better auto-complete while editing the extension.
// #define EDIT_EXTENSION

namespace rerun {
    namespace components {

#ifdef EDIT_EXTENSION
        struct Rotation3DExt {
            Rotation3D repr;
#define Rotation3D Rotation3DExt

            // <CODEGEN_COPY_TO_HEADER>

            static const Rotation3D IDENTITY;

            /// Construct Rotation3d from Quaternion.
            Rotation3D(datatypes::Quaternion quaternion) : repr{quaternion} {}

            /// Construct Rotation3d from axis-angle
            Rotation3D(datatypes::RotationAxisAngle axis_angle) : repr{axis_angle} {}

            // </CODEGEN_COPY_TO_HEADER>
        };

#undef Rotation3DExt
#else
#define Rotation3DExt Rotation3D
#endif

        const Rotation3DExt Rotation3DExt::IDENTITY = datatypes::Quaternion::IDENTITY;
    } // namespace components
} // namespace rerun
